src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeWriter.java

```diff
@@ -0,0 +1,779 @@
+                              .toString();
+        }
+    }
+
+    private class TreeWriter
+    {
+        private final IndexOutput treeOutput;
+        private final List<Long> leafBlockFilePointer = new ArrayList<>();
+        private final List<byte[]> leafBlockStartValues = new ArrayList<>();
+        private final byte[] leafValues = new byte[maxPointsInLeafNode * bytesPerValue];
+        private final long[] leafRowIDs = new long[maxPointsInLeafNode];
+        private final RowIDAndIndex[] rowIDAndIndexes = new RowIDAndIndex[maxPointsInLeafNode];
+        private final int[] orderIndex = new int[maxPointsInLeafNode];
+        private final Callback callback;
+        private final GrowableByteArrayDataOutput scratchOut = new GrowableByteArrayDataOutput(32 * 1024);
+        private final GrowableByteArrayDataOutput scratchOut2 = new GrowableByteArrayDataOutput(2 * 1024);
+        private final byte[] lastPackedValue = new byte[bytesPerValue];
+
+        private long valueCount;
+        private int leafValueCount;
+        private long lastRowID;
+
+        TreeWriter(IndexOutput treeOutput, Callback callback)
+        {
+            assert callback != null : "Callback cannot be null in TreeWriter";
+
+            this.treeOutput = treeOutput;
+            this.callback = callback;
+
+            for (int x = 0; x < rowIDAndIndexes.length; x++)
+            {
+                rowIDAndIndexes[x] = new RowIDAndIndex();
+            }
+        }
+
+        void add(byte[] packedValue, long docID) throws IOException
+        {
+            if (DEBUG)
+                valueInOrder(valueCount + leafValueCount, lastPackedValue, packedValue, 0, docID, lastRowID);
+
+            System.arraycopy(packedValue, 0, leafValues, leafValueCount * bytesPerValue, bytesPerValue);
+            leafRowIDs[leafValueCount] = docID;
+            leafValueCount++;
+
+            if (leafValueCount == maxPointsInLeafNode)
+            {
+                // We write a block once we hit exactly the max count
+                writeLeafBlock();
+                leafValueCount = 0;
+            }
+
+            if (DEBUG)
+                if ((lastRowID = docID) < 0)
+                    throw new AssertionError("document id must be >= 0; got " + docID);
+        }
+
+        /**
+         * Write a leaf block if we have unwritten values and return the total number of values added
+         */
+        public long finish() throws IOException
+        {
+            if (leafValueCount > 0)
+                writeLeafBlock();
+
+            return valueCount;
+        }
+
+        private void writeLeafBlock() throws IOException
+        {
+            assert leafValueCount != 0;
+            if (valueCount == 0)
+            {
+                System.arraycopy(leafValues, 0, minPackedValue, 0, bytesPerValue);
+            }
+            System.arraycopy(leafValues, (leafValueCount - 1) * bytesPerValue, maxPackedValue, 0, bytesPerValue);
```

Comment:
Note: We only write the min value on the first block, and subsequently overwrite the max on every block.

(No action necessary on these...just making notes to myself that may or may not become inline comments...)